background/usercss-helper.js

```diff
@@ -1,42 +1,66 @@
-/* global API_METHODS usercss chromeLocal styleManager FIREFOX deepCopy openURL
-  download */
+/* global API_METHODS usercss styleManager deepCopy openURL download URLS getTab */
+/* exports usercssHelper */
 'use strict';
 
-(() => {
+// eslint-disable-next-line no-unused-vars
+const usercssHelper = (() => {
+  const installCodeCache = {};
+  const clearInstallCode = url => delete installCodeCache[url];
+  const isResponseText = r => /^text\/(css|plain)(;.*?)?$/i.test(r.headers.get('content-type'));
+  // detecting FF68 by the added feature as navigator.ua may be spoofed via about:config or devtools
+  const fileLoader = !chrome.app && chrome.storage.managed &&
+    (tabId => browser.tabs.executeScript(tabId, {file: '/content/download-self.js'}).then(r => r[0]));
+
   API_METHODS.installUsercss = installUsercss;
   API_METHODS.editSaveUsercss = editSaveUsercss;
   API_METHODS.configUsercssVars = configUsercssVars;
 
   API_METHODS.buildUsercss = build;
-  API_METHODS.openUsercssInstallPage = install;
-
   API_METHODS.findUsercss = find;
 
-  const TEMP_CODE_PREFIX = 'tempUsercssCode';
-  const TEMP_CODE_CLEANUP_DELAY = 60e3;
-  let tempCodeLastWriteDate = 0;
-  if (FIREFOX) {
-    // the temp code is created on direct installation of usercss URLs in FF
-    // and can be left behind in case the install page didn't open in time before
-    // the extension was updated/reloaded/disabled or the browser was closed
-    setTimeout(function poll() {
-      if (Date.now() - tempCodeLastWriteDate < TEMP_CODE_CLEANUP_DELAY) {
-        setTimeout(poll, TEMP_CODE_CLEANUP_DELAY);
-        return;
+  API_METHODS.getUsercssInstallCode = url => {
+    // when the installer tab is reloaded after the cache is expired, this will throw intentionally
+    const {code, timer} = installCodeCache[url];
+    clearInstallCode(url);
+    clearTimeout(timer);
+    return code;
+  };
+
+  return {
+
+    testUrl(url) {
+      return url.includes('.user.') &&
+        /^(https?|file|ftps?):/.test(url) &&
+        /\.user\.(css|styl)$/.test(url.split(/[#?]/, 1)[0]);
+    },
+
+    /** @return {Promise<{ code:string, inTab:boolean } | false>} */
+    testContents(tabId, url) {
+      const isFile = url.startsWith('file:');
+      const inTab = isFile && Boolean(fileLoader);
+      return Promise.resolve(isFile || fetch(url, {method: 'HEAD'}).then(isResponseText))
+        .then(ok => ok && (inTab ? fileLoader(tabId) : download(url)))
+        .then(code => /==userstyle==/i.test(code) && {code, inTab});
```

Comment:
The function may return `false`. It should be:
```js
.then(code => ({
  code: code && /==userstyle==/i.test(code) ? code : null,
  inTab
}));
```

Reply:
Nope, it's perfectly fine to use test() directly because it stringifies its argument. This is how JavaScript works, it's not C# or python or Java.

Reply:
Promise-based code is inherently awkward and clumsy so the real solution is to switch to async/await:

```js
    async testContents(tabId, url) {
      const isFile = url.startsWith('file:');
      const inTab = isFile && Boolean(fileLoader);
      if ((isFile || isResponseText(await fetch(url, {method: 'HEAD'}))) &&
          /==userstyle==/i.test(await inTab ? fileLoader(tabId) : download(url))) {
        return {code, inTab};
      }
    },
```

Reply:
What is the signature of `testContents`? If you want to allow the function to return `null/undefined` then the code using it needs to be changed:
https://github.com/openstyles/stylus/blob/01873ae8e8f7374b141fdb921100ecc53d1a78ab/background/background.js#L94-L96

----

To write clean callback code, you can refactor phases into functions:
```js
testContents(tabId, url) {
  const isFile = url.startsWith('file:');
  const inTab = isFile && Boolean(fileLoader);
  return testSourceContent().then(ok => {
    if (ok) {
      return downloadCode().then(code => ({
        code: /==userstyle==/i.test(code) ? code : null,
        inTab
      }));
    }
  });
  
  function testSourceContent() {
    return isFile ? Promise.resolve(true) :
      fetch(url, {method: 'HEAD'}).then(isResponseText);
  }
  
  function downloadCode() {
    return inTab ? fileLoader(tabId) : download(url);
  }
}
```

Reply:
1) The signature is given in the JSDoc comment, start using it already. 2) Your example code is only superficially clean because it simply adds a few empty lines and gives names to intermediate things, the same can be achieved by adding comments in the original code. I often use such separation myself but it's not needed here because the code is already trivial, no need to nitpick on it.